---
title: "MOFA results plotting and enrichment"
author: "Scott Ginebaugh"
date: "2023-05-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.width = 7, fig.height = 7, cache = FALSE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

```{r load_libraries, include = FALSE, message = FALSE, warning = FALSE}
library(Seurat)
library(qs)
library(magrittr)
library(dplyr)
library(tidyverse)
library(MOFA2)
library(MOFAcellulaR)

library(ggplot2)
library(ggpmisc)
library(ggprism)
library(knitr)
```

```{r functions, include = FALSE, message = FALSE, warning = FALSE}
## TODO move functions to a function file and pull in
getFactorLoadings <- function(model, factor) {
  factor_loadings <- get_weights(model, as.data.frame = T) %>%
    as.data.frame() %>%
    dplyr::mutate(
      feature = strsplit(as.character(feature), "_") %>%
        map_chr(., ~ .x[[2]]),
      ctype = strsplit(as.character(view), "_") %>%
        map_chr(., ~ .x[[1]])
    ) %>%
    dplyr::rename("factors" = factor) %>%
    dplyr::select(-view) %>%
    dplyr::filter(factors == factor) %>%
    dplyr::select(-factors)

  return(factor_loadings)
}


## code taken mostly from scTID paper
gseAnalysis <- function(geneList,Annotation_DB){
  
  geneList = geneList[geneList %in% unique(unlist(Annotation_DB))]
  
  ResultsDF = matrix(0,nrow = length(Annotation_DB),ncol = 5)
  rownames(ResultsDF) = names(Annotation_DB)
  colnames(ResultsDF) = c("GenesInPathway","GenesInList","GeneNames","p_value","corr_p_value")
  
  DB_genecontent = length(unique(unlist(Annotation_DB)))
  GenesDB = DB_genecontent 
  SelectedGenes = length(geneList)
  
  for(gset in rownames(ResultsDF)){
    GP = length(Annotation_DB[[gset]])
    GL = length(intersect(Annotation_DB[[gset]],geneList))
    
    ResultsDF[gset,"GenesInList"] = GL
    ResultsDF[gset,"GenesInPathway"] = GP
    ResultsDF[gset,"GeneNames"] = paste(intersect(Annotation_DB[[gset]],geneList),collapse = ",")
    ResultsDF[gset,"p_value"] = phyper(q=GL - 1, m=GP, n=GenesDB-GP, k=SelectedGenes, lower.tail = FALSE, log.p = FALSE)
  }
  
  ResultsDF[,"corr_p_value"] = p.adjust(ResultsDF[,"p_value"],method = "BH")
  ResultsDF = data.frame(ResultsDF,stringsAsFactors = F)
  ResultsDF = ResultsDF[order(ResultsDF[,"p_value"]),]
  
  ResultsDF = ResultsDF %>% 
    rownames_to_column("gset") %>% 
    mutate_at(c("GenesInPathway","GenesInList",
                "p_value","corr_p_value"), 
              as.numeric) %>% 
    dplyr::arrange(corr_p_value,GenesInList)
  
  return(ResultsDF)
}
```

```{r import_data, include = FALSE, message = FALSE, warning = FALSE}
seurat <- qread("data/processed/annotated_seurat_perez.qs")

mofa_model <- qread("data/processed/mofa_model_perez.qs")

sample_meta <- unique(seurat@meta.data[, c("sample", "age_group", "age_numeric")])

```

# Introduction

A common method to identify potential therapeutic targets in transcriptomic data is to use differential expression (DE) to identify highly differentially expressed genes between two conditions. When it comes to single-cell/single-nucleus datasets, DE is great at identifying transcriptomic differences between cell types (a method commonly used to identify marker genes). ScRNAseq is also very useful for identifying differences in cell type quantities between samples/conditions. However, DE is often not ideal for identifying within-cluster, between-sample/condition transcriptomic changes in scRNAseq datasets (e.g., what are the transcriptomic differences between muscle endothelial cells from young and old individuals). This is because batch effects, noise, variation in the number of cells per sample, and variation in the number of samples/donors per group are often quite large in scRNAseq, leading to in inaccurate results from DE.

Cells can have both cell-type specific and shared responses to the same cue, and well as feedback loops that occur through cell-cell interactions and between-organ signals (e.g. hormones). These multicellular changes are sometimes refered to as "multicellular programs". Between-sample DE looks at one cell-type at a time and does not consider these multicellular programs, which can result in "missing the forest for the trees". This is especially true for aging, which clearly impacts every celltype in the body in numerous ways.

We need more sophisticated methods that are able to extract important signals from noise, and identify both celltype-specific and multicellular program changes. One such approach is to use Blind Source Separation techniques, such as PCA, ICA, and NMF. However, when applied directly to scRNAseq data, these methods will mainly identify differences between cell types. This limitation can be overcome by taking our genes x cells matrix, and turn it into a pseudobulk celltype x genes x samples tensor. We can then decompose the tensor with blind source separation to pull out the important biological signals.

Here, we analyze single-nucleus RNAseq data from Perez et al. 2022 (https://doi.org/10.18632/aging.204435) using MOFAcell (https://doi.org/10.7554/eLife.93161) -- a factor analysis approach to single cell data. Ideally, this method will identify a factor that is highly correlated with age. That "Aging Factor" should then contain a scored list of how much each gene contributes to aging in each cell type, independent of technical noise or other biological factors.


```{r prepare_inputs, include = FALSE, message = FALSE, warning = FALSE}
seurat <- qread("data/processed/annotated_seurat_perez.qs")

mofa_model <- qread("data/processed/mofa_model_perez.qs")

sample_meta <- unique(seurat@meta.data[, c("sample", "age_group", "age_numeric")])

## set Aging Factor
aging_factor <- "Factor4"

```

# Results

## Dataset overview

This dataset contains n=17 biopsies from the vastus lateralis of 6 younger and 11 older individuals. In the original analysis, they used the CellRanger filtering to identify empty droplets. However, Cellranger tends to be underpowered for single-nucleus data. Furthermore, earlier versions of Cellranger (which were likely used based on the publication date of the manuscript) did a poor job of handling single nucleus data. 

To resolve this, I downloaded the original fastq files from SRA, realigned them with Cellranger v7.1.0, and then used Cellbender to select empty droplets. However, I used the counts from Cellranger, rather than the RNA-decontaminated Cellbender counts, because I suspect that samples from old donors likely have more noise in their transcriptome due to age-related loss of cellular epigenetic information, and I was concerned that ambient RNA removal may accidently remove important biological signal.

After alignment, filtering, batch correction, and annotation, the final dataset consisted of 176,861 cells -- over 30,000 more cells than the original manuscript, even though I used more stringent cell QC filtering than the original manuscript. This also allows us to identify two distinct groups of neural NMJ populations (schwann cells), as well as a distinct group of mast cells, that were not identified in the original manuscript.

### Dataset UMAP {.tabset}

#### Cell types

```{r umap_celltype, echo = FALSE, message = FALSE, warning = FALSE}
DimPlot(seurat, group.by = "annotation_level1", label = TRUE)
```

#### Age Group

```{r umap_agegroup, echo = FALSE, message = FALSE, warning = FALSE}
DimPlot(seurat, group.by = "age_group", label = FALSE)
```

#### Sample

```{r umap_sample, echo = FALSE, message = FALSE, warning = FALSE}
DimPlot(seurat, group.by = "sample", label = FALSE)
```

#### Phase

```{r umap_phase, echo = FALSE, message = FALSE, warning = FALSE}
DimPlot(seurat, group.by = "sample", label = FALSE)
```

#### QC

```{r umap_qc, echo = FALSE, message = FALSE, warning = FALSE}
FeaturePlot(seurat, min.cutoff = "q1", max.cutoff = "q99",
            features = c("nCount_RNA","nFeature_RNA","mitoRatio","riboRatio"))
```

### {-}

## Identification of Aging Factor

We construct the data tensor, then decompose it into 4 factors using MOFAcell. We see that factor 4 is strongly correlated with aging.

```{r prep_heatmap, echo = FALSE, message = FALSE, warning = FALSE}
## matrix of gene contributions from each celltype to factor
factor_loadings_mat <- getFactorLoadings(model = mofa_model, factor = aging_factor) %>%
  pivot_wider(names_from = ctype, values_from = value, values_fill = 0) %>%
  column_to_rownames("feature") %>%
  as.matrix()


## sample expression of each factor in each sample, for heatmaps
all_factors <- MOFAcellulaR::get_tidy_factors(
  model = mofa_model,
  metadata = sample_meta,
  factor = "all",
  sample_id_column = "sample"
)

## association of each factor with clinical info of interest (e.g. aging)
expl_var_pgroup <- MOFAcellulaR::get_associations(
  model = mofa_model,
  metadata = sample_meta,
  sample_id_column = "sample",
  test_variable = "age_group",
  test_type = "categorical",
  group = FALSE
)


## -------------------------heatmap setup-----------------------------------
## global plotting settings for MOFA2 package
ht_opt$ROW_ANNO_PADDING <- unit(2.5, "mm")
ht_opt$COLUMN_ANNO_PADDING <- unit(2.5, "mm")
## Association colors
col_fun_assoc <- circlize::colorRamp2(seq(0, 5, length = 20), hcl.colors(20, "Purples", rev = T))
col_fun_r2 <- circlize::colorRamp2(seq(0, 100, length = 50), hcl.colors(50, "Oranges", rev = T))
## Make the heatmap annotation similar to scITD
row_anns <- do.call(rbind, get_factors(mofa_model, factors = "all",groups = "all")) %>%
  rownames() %>%
  enframe() %>%
  dplyr::select(value) %>%
  left_join(sample_meta, by = c("value" = "sample")) %>%
  dplyr::select(age_group)


row_ha <- rowAnnotation(
  age = row_anns$age_group,
  gap = unit(2.5, "mm"),
  border = TRUE,
  col = list(age = c("Young" = "green", "Old" = "darkred"))
)


assoc_pvals <- bind_rows("age_group" = -log10(expl_var_pgroup$adj_pvalue)) %>%
  as.matrix()
## Add explain variance per cell-type per group
r2_per_factor <- list()
r2_per_factor <- do.call(cbind, r2_per_factor)

column_ha <- HeatmapAnnotation(
  "r2" = r2_per_factor,
  "assocs" = assoc_pvals,
  gap = unit(2.5, "mm"),
  border = TRUE,
  col = list(r2 = col_fun_r2, assocs = col_fun_assoc)
)


factor_score_mat <- do.call(rbind,get_factors(mofa_model, factors = "all", groups = "all"))

```

```{r plot_heatmap1, echo = FALSE, message = FALSE, warning = FALSE}
## all factors
col_fun_fact_all <- circlize::colorRamp2(
  seq(
    -max(abs(factor_score_mat)),
    max(abs(factor_score_mat)),
    length = 50
  ),
  hcl.colors(50, "Green-Brown", rev = TRUE)
)

scores_hmap <- Heatmap(
  factor_score_mat,
  name = "factor_scores",
  right_annotation = row_ha,
  top_annotation = column_ha,
  cluster_columns = FALSE,
  show_row_dend = TRUE,
  show_row_names = FALSE,
  border = TRUE,
  gap = unit(2.5, "mm"),
  col = col_fun_fact_all
)
scores_hmap

```

```{r first_treatment_table, echo = FALSE, message = FALSE, warning = FALSE}
expl_var_pgroup %>% 
  kable(caption = "Difference in Factor Score Between Young and Old")

```
```{r set_aging_factor, echo = FALSE, message = FALSE, warning = FALSE}
aging_factor <- "Factor4"
```

However, we also notice that factor 1 is strongly expressed in several samples (to the point where it messes up the color scaling on our heatmap), and is not associated with aging. Further investigation into the expression pattern of Factor 1 reveals that it is strongly correlated with poor sample quality. This is a good result, as it shows that our approach is able to separate technical noise from meaningful biological signals 

```{r factor1_cor, echo = FALSE, message = FALSE, warning = FALSE}

## scatter plot of mean mitochondrial ratio v noise factor 
mean_sample_qc_scores <- seurat@meta.data[, c("sample", "mitoRatio", "nCount_RNA", "nFeature_RNA", "riboRatio", "log10GenesPerUMI")] %>%
  group_by(sample) %>%
  summarise_all("mean") %>%
  left_join(
    .,
    rownames_to_column(data.frame(factor_score_mat), var = "sample"),
    by = "sample"
  ) %>%
  select(-sample)

mean_sample_qc_scores %>%
  ggplot(aes(x = Factor1, y = mitoRatio)) +
  geom_point() +
  stat_poly_eq(use_label("R2","P"), size = 5) +
  stat_poly_line(se = FALSE) +
  theme_prism() +
  theme(aspect.ratio = 1) +
  ggtitle("Mean Mitochondrial Ratio vs Factor 1 Score")

mean_sample_qc_scores %>%
  ggplot(aes(x = Factor1, y = nCount_RNA)) +
  geom_point() +
  stat_poly_eq(use_label("R2","P"), size = 5, label.x = "right" ) +
  stat_poly_line(se = FALSE) +
  theme_prism() +
  theme(aspect.ratio = 1) +
  ggtitle("Mean Gene Expression Count vs Factor 1 Score")
```

When we remove Factor 1 from our heatmap, we see a nice separation of young and old samples.

```{r plot_heatmap2, echo = FALSE, message = FALSE, warning = FALSE}
## remove factor one because it is noise
col_fun_fact_nonoise <- circlize::colorRamp2(
  seq(
    -max(abs(factor_score_mat[,-1])),
    max(abs(factor_score_mat[,-1])),
    length = 50
  ),
  hcl.colors(50, "Green-Brown", rev = TRUE)
)
Heatmap(
  factor_score_mat[, -1],
  name = "factor_scores",
  right_annotation = row_ha,
  cluster_columns = FALSE,
  show_row_dend = TRUE,
  show_row_names = FALSE,
  border = TRUE,
  gap = unit(2.5, "mm"),
  col = col_fun_fact_nonoise
)

```

In fact, Factor 4 perfectly separates young from old donors. At no point was the algorithm ever informed of the donor groupings, so this shows that this approach can extract meaningful biological signals in an unsupervised manor.

```{r plot_agefactor_boxplot, echo = FALSE, message = FALSE, warning = FALSE}
group_boxplot <- all_factors %>%
  dplyr::filter(Factor %in% c(aging_factor)) %>%
  ggplot(aes(x = age_group, y = value, fill = age_group)) +
  geom_boxplot() +
  geom_point() +
  theme_prism() +
  scale_fill_manual(values = c("darkred", "green")) +
  theme(
    axis.text = element_text(size = 12),
    axis.title = element_text(size = 12),
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
  ) +
  ylab("Factor 4 score") +
  xlab("Age Group")
plot(group_boxplot)


```

## Analysis of Aging Factor

Now that we have identified the Aging Factor, and shown that it is independent of technical noise, we will analyze the factor to extract biological insights. First, we can investigate how much each celltype contributes to the Aging Factor. We see that the skeletal muscle cells are the strongest contributors, followed by the schwann cells. This is not surprising, since loss of neurotransmission to muscle cells is a major factor in muscle wasting (in fact, crushing the motor neurons innervating a muscle is a common method to create mouse models of muscle wasting)

```{r celltype_contributions, echo = FALSE, message = FALSE, warning = FALSE}
mofa_model@cache$variance_explained$r2_per_factor$single_group[4, ] %>%
  enframe() %>%
  ggplot(aes(x = name, y = value)) +
  geom_bar(stat = "identity") +
  theme_prism() +
  theme(
    axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5),
    axis.text = element_text(size = 12)
  ) +
  ylab("R2") +
  xlab("") 
```

Next, we look at the the gene contributions to the Aging Factor in each celltype. We see that a large majority of genes only contribute to the Aging Factor in a few cell types. However, there about 14% of genes contribute to the Aging Factor in more than 3 celltypes. Further analysis of these genes could identify high-value targets to reverse age-related muscle wasting (but is not included in this analysis here).

```{r gene_contribution_heatmap, echo = FALSE, message = FALSE, warning = FALSE}
Heatmap(factor_loadings_mat, show_row_names = FALSE)

factor_loadings_filt <- getFactorLoadings(model = mofa_model, factor = aging_factor) %>%
  dplyr::filter(abs(value) >= 0.3)

shared_genes <- factor_loadings_filt %>%
  group_by(feature) %>%
  summarize(n_cells = length(feature)) %>%
  arrange(-n_cells) %>%
  group_by(n_cells) %>%
  summarize(count = length(n_cells))

shared_genes_plt <- ggplot(shared_genes, aes(x = (count), y = reorder(paste(n_cells, "types"), n_cells))) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  theme(
    axis.text.y = element_text(size = 12),
    axis.text.x = element_text(size = 12)
  ) +
  theme(plot.margin = unit(c(0, 1, 0.2, 0.2), "cm")) +
  ylab("number of cell types") +
  xlab("number of genes")
shared_genes_plt
```

Next, we perform pathway enrichment analysis of the Aging Factor
















